src/input/keyboard/mod.rs

```diff
@@ -130,15 +130,61 @@ impl LedState {
     }
 }
 
+/// An xkbcommon context, keymap, and state, that can be sent to another
+/// thread, but should not have additional ref-counts kept on one thread.
+pub struct Xkb {
+    context: xkb::Context,
+    keymap: xkb::Keymap,
+    state: xkb::State,
+}
+
+impl Xkb {
+    /// The xkbcommon context.
+    ///
+    /// # Safety
+    /// A ref-count of the context should not outlive the `Xkb`
+    pub unsafe fn context(&self) -> &xkb::Context {
+        &self.context
+    }
+
+    /// The xkbcommon keymap.
+    ///
+    /// # Safety
+    /// A ref-count of the keymap should not outlive the `Xkb`
+    pub unsafe fn keymap(&self) -> &xkb::Keymap {
+        &self.keymap
+    }
+
+    /// The xkbcommon state.
+    ///
+    /// # Safety
+    /// A ref-count of the state should not outlive the `Xkb`
+    pub unsafe fn state(&self) -> &xkb::State {
+        &self.state
+    }
+}
+
+impl fmt::Debug for Xkb {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("Xkb")
+            .field("context", &self.context.get_raw_ptr())
+            .field("keymap", &self.keymap.get_raw_ptr())
+            .field("state", &self.state.get_raw_ptr())
+            .finish()
+    }
+}
+
+// This is OK because all parts of `xkb` will remain on the
+// same thread
+unsafe impl Send for Xkb {}
```

Comment:
But `Xkb` is part of smithay's public api right? So this is only safe, because all the accessors aren't?

In that case wouldn't it make more sense to wrap this in a way, where we don't expose the safety issues to the outside?

Something like:

```rust
pub struct Xkb(XkbInternal);

struct XkbInternal {
    context: xkb::Context,
    keymap: xkb::Keymap,
    state: xkb::State,
}

unsafe impl Send for XkbInternal {}
```

Reply:
I think we can leave it with accessors in that case.